File: Source/Aladdin.CAPI.PKCS12.Java/src/aladdin/capi/pkcs12/Pfx.java
package aladdin.capi.pkcs12;
import aladdin.asn1.*;
import aladdin.asn1.Integer; 
import aladdin.asn1.iso.*;
import aladdin.asn1.iso.pkcs.*;
import aladdin.asn1.iso.pkcs.pkcs7.*;
import aladdin.asn1.iso.pkcs.pkcs8.*;
import aladdin.asn1.iso.pkcs.pkcs12.*;
import aladdin.capi.*;
import aladdin.capi.SecretKey;
import aladdin.capi.Certificate;
import aladdin.capi.pbe.*;
import java.security.*;
import java.io.*;
import java.util.*;

public final class Pfx 
{
	///////////////////////////////////////////////////////////////////////
    // Создать контейнер
	///////////////////////////////////////////////////////////////////////
    public static PFX createContainer( 
        Factory factory, AuthenticatedSafe authenticatedSafe) throws IOException
    { 
        // указать идентификатор данных
        String dataType = aladdin.asn1.iso.pkcs.pkcs7.OID.DATA; 
        
		// закодировать данные
		byte[] encoded = authenticatedSafe.encoded();  

        // закодировать данные
        ContentInfo contentInfo = new ContentInfo(
            new ObjectIdentifier(dataType), new OctetString(encoded)
        ); 
        // закодировать контейнер
        return new PFX(new Integer(3), contentInfo, null); 
    }
	///////////////////////////////////////////////////////////////////////
	// Создать подписанный контейнер PKCS12
	///////////////////////////////////////////////////////////////////////
	public static PFX createSignedContainer(Factory factory, IRand rand,  
        Integer version, AuthenticatedSafe authenticatedSafe, 
		IPrivateKey privateKey, Certificate certificate, 
		AlgorithmIdentifier hashParameters, 
		AlgorithmIdentifier signParameters, 
		Attributes authAttributes, Attributes unauthAttributes, 
		CertificateSet certificates, RevocationInfoChoices crls) throws IOException
	{
		// указать тип данных
		String dataType = aladdin.asn1.iso.pkcs.pkcs7.OID.DATA; 

		// указать алгоритмы хэширования
		AlgorithmIdentifiers digestAlgorithms = new AlgorithmIdentifiers(
            new AlgorithmIdentifier[] {hashParameters}
        ); 
        // создать вложенные данные
		EncapsulatedContentInfo encapContentInfo = 
			new EncapsulatedContentInfo(
				new ObjectIdentifier(dataType), 
                new OctetString(authenticatedSafe.encoded())
		); 
		// подписать данные
		SignerInfo signerInfo = CMS.signData(rand, 
			privateKey, certificate, hashParameters, signParameters, 
            encapContentInfo, authAttributes, unauthAttributes
		); 
		// закодировать зашифрованные данные
		SignerInfos signerInfos = new SignerInfos(new SignerInfo[] {signerInfo});
        
		// закодировать структуру CMS
		SignedData signedData = new SignedData(version, 
			digestAlgorithms, encapContentInfo, certificates, crls, signerInfos
		);
        // указать идентификатор данных
        String oid = aladdin.asn1.iso.pkcs.pkcs7.OID.SIGNED_DATA; 
        
		// закодировать данные
		ContentInfo contentInfo = new ContentInfo(
                new ObjectIdentifier(oid), signedData
		); 
        // закодировать контейнер
		return new PFX(new Integer(3), contentInfo, null); 
	}
	///////////////////////////////////////////////////////////////////////
	// Создать контейнер PKCS12 с вычисленной имитовставкой
	///////////////////////////////////////////////////////////////////////
	public static PFX createAuthenticatedContainer(Factory factory, 
        AuthenticatedSafe authenticatedSafe, AlgorithmIdentifier hashParameters, 
        byte[] salt, int iterations, String password) throws IOException
	{
        // закодировать данные
        byte[] encoded = authenticatedSafe.encoded();  

		// создать алгоритм хэширования 
		try (Hash hashAlgorithm = (Hash)factory.createAlgorithm(null, hashParameters, Hash.class))
        {
            // при ошибке выбросить исключение
            if (hashAlgorithm == null) throw new UnsupportedOperationException(); 

            // создать алгоритм вычисления имитовставки
            try (Mac macAlgorithm = new PBMACP12(hashAlgorithm, salt, iterations)) 
            {
                // закодировать пароль
                try (ISecretKey encodedPassword = SecretKey.fromPassword(password, "UTF-8")) 
                {
                    // вычислить имитовставку от данных
                    OctetString mac = new OctetString(macAlgorithm.macData(
                        encodedPassword, encoded, 0, encoded.length
                    )); 
                    // закодировать имитовставку от данных
                    DigestInfo digestInfo = new DigestInfo(hashParameters, mac); 

                    // закодировать информацию об имитовставке
                    MacData macData = new MacData(digestInfo, 
                        new OctetString(salt), new Integer(iterations)
                    ); 
                    // указать идентификатор данных
                    String oid = aladdin.asn1.iso.pkcs.pkcs7.OID.DATA; 

                    // закодировать данные
                    ContentInfo contentInfo = new ContentInfo(
                        new ObjectIdentifier(oid), new OctetString(encoded)
                    ); 
                    // закодировать контейнер
                    return new PFX(new Integer(3), contentInfo, macData); 
                }
                // обработать неожидаемое исключение
                catch (InvalidKeyException e) { throw new RuntimeException(e); }
            }
        }
    }
	///////////////////////////////////////////////////////////////////////
	// Проверить корректность контейнера PKCS12
	///////////////////////////////////////////////////////////////////////
	public static void checkAuthenticatedContainer(Factory factory, 
        PFX container, String password) throws IOException
    {
		// получить закодированную имитовставку
		MacData macData = container.macData(); byte[] salt = macData.macSalt().value(); 

		// получить число итераций
		int iterations = macData.iterations().value().intValue(); 

        // получить имитовставку
        byte[] mac = macData.mac().digest().value(); 
        
        // получить проверяемые данные
        AuthenticatedSafe authenticatedSafe = container.getAuthSafeContent(); 

		// закодировать данные
		byte[] encoded = authenticatedSafe.encoded();  
        
		// создать алгоритм хэширования 
		try (Hash hashAlgorithm = (Hash)factory.createAlgorithm(null, macData.mac().digestAlgorithm(), Hash.class))
        {
            // при ошибке выбросить исключение
            if (hashAlgorithm == null) throw new UnsupportedOperationException(); 

            // создать алгоритм вычисления имитовставки
            try (Mac macAlgorithm = new PBMACP12(hashAlgorithm, salt, iterations)) 
            {
                // закодировать пароль
                try (ISecretKey encodedPassword = SecretKey.fromPassword(password, "UTF-8")) 
                {
                    // вычислить имитовставку от данных
                    byte[] check = macAlgorithm.macData(encodedPassword, encoded, 0, encoded.length);

                    // проверить совпадение имитовставок
                    if (!Arrays.equals(check, mac)) 
                    {
                        // создать алгоритм вычисления имитовставки
                        try (Mac macAlgorithm2 = new PBMACTС26(hashAlgorithm, salt, iterations)) 
                        {
                            // вычислить имитовставку от данных
                            check = macAlgorithm2.macData(encodedPassword, encoded, 0, encoded.length);
                        }
                    }
                    // проверить совпадение имитовставок
                    if (!Arrays.equals(check, mac)) throw new AuthenticationException();
                }
                // обработать неожидаемое исключение
                catch (InvalidKeyException e) { throw new RuntimeException(e); }
            }
        }
    }
	///////////////////////////////////////////////////////////////////////
	// Зашифровать элемент 
	///////////////////////////////////////////////////////////////////////
	static SafeBag encrypt(PfxData<SafeBag> safeBag) throws IOException
	{
		// проверить наличие шифрования
		if (safeBag.encryptor == null) return safeBag.content; 

		// в зависимости от типа
		if (safeBag.content.bagId().value().equals(aladdin.asn1.iso.pkcs.pkcs12.OID.BT_KEY))
		{
			// раскодировать тип данных
			PrivateKeyInfo keyInfo = new PrivateKeyInfo(safeBag.content.bagValue());

			// зашифровать данные
			IEncodable encodable = Encodable.decode(
                safeBag.encryptor.encrypt(keyInfo.encoded())
            ); 
			// преобразовать тип данных
			EncryptedPrivateKeyInfo encryptedKey = new EncryptedPrivateKeyInfo(encodable);

			// вернуть зашифрованное представление
			return new SafeBag(
                new ObjectIdentifier(aladdin.asn1.iso.pkcs.pkcs12.OID.BT_SHROUNDED_KEY), 
                encryptedKey, safeBag.content.bagAttributes()
            );
		}
		else return safeBag.content; 
	}
	///////////////////////////////////////////////////////////////////////
	// Расшифровать элемент
	///////////////////////////////////////////////////////////////////////
	static PfxData<SafeBag> decrypt(SafeBag safeBag, PfxDecryptor decryptor) throws IOException
	{
		// для незашифрованного элемента
		if (!safeBag.bagId().value().equals(aladdin.asn1.iso.pkcs.pkcs12.OID.BT_SHROUNDED_KEY))
		{
			// вернуть исходное содержимое
			return new PfxData<SafeBag>(safeBag, null);
		}
		// расшифровать данные
		PfxData<byte[]> decrypted = decryptor.decrypt(
            safeBag.bagValue().encoded(), EncryptedPrivateKeyInfo.class
        );
		// преобразовать тип данных 
		PrivateKeyInfo keyInfo = new PrivateKeyInfo(Encodable.decode(decrypted.content));

        // вернуть расшифрованные данные
		safeBag = new SafeBag(
            new ObjectIdentifier(aladdin.asn1.iso.pkcs.pkcs12.OID.BT_KEY), 
            keyInfo, safeBag.bagAttributes()
        );
        // вернуть расшифрованные данные
        return new PfxData<SafeBag>(safeBag, decrypted.encryptor); 
	}
}

File: Source/Aladdin.CAPI.PKCS12/CAPI/PKCS12/Pfx.cs
using System;
using System.Text;
using System.Security.Authentication; 

namespace Aladdin.CAPI.PKCS12
{
	public static class Pfx
	{
	    ///////////////////////////////////////////////////////////////////////
        // Создать контейнер
	    ///////////////////////////////////////////////////////////////////////
        public static ASN1.ISO.PKCS.PKCS12.PFX CreateContainer( 
            Factory factory, ASN1.ISO.PKCS.PKCS12.AuthenticatedSafe authenticatedSafe) 
        { 
            // указать идентификатор данных
			string dataType = ASN1.ISO.PKCS.PKCS7.OID.data; 
        
		    // закодировать данные
		    byte[] encoded = authenticatedSafe.Encoded;  

            // закодировать данные
            ASN1.ISO.PKCS.ContentInfo contentInfo = new ASN1.ISO.PKCS.ContentInfo(
                new ASN1.ObjectIdentifier(dataType), new ASN1.OctetString(encoded)
            ); 
            // закодировать контейнер
            return new ASN1.ISO.PKCS.PKCS12.PFX(new ASN1.Integer(3), contentInfo, null); 
        }
		///////////////////////////////////////////////////////////////////////
		// Создать подписанный контейнер PKCS12
		///////////////////////////////////////////////////////////////////////
		public static ASN1.ISO.PKCS.PKCS12.PFX CreateSignedContainer(
			Factory factory, IRand rand, ASN1.Integer version,
			ASN1.ISO.PKCS.PKCS12.AuthenticatedSafe authenticatedSafe,  
			IPrivateKey privateKey, Certificate certificate, 
			ASN1.ISO.AlgorithmIdentifier hashParameters, 
			ASN1.ISO.AlgorithmIdentifier signParameters, 
			ASN1.ISO.Attributes authAttributes, ASN1.ISO.Attributes unauthAttributes, 
			ASN1.ISO.CertificateSet certificates, ASN1.ISO.RevocationInfoChoices crls)
		{
			// указать тип данных
			string dataType = ASN1.ISO.PKCS.PKCS7.OID.data; 

			// указать алгоритмы хэширования
			ASN1.ISO.AlgorithmIdentifiers digestAlgorithms = 
				new ASN1.ISO.AlgorithmIdentifiers(
                    new ASN1.ISO.AlgorithmIdentifier[] { hashParameters }
            ); 
			// создать вложенные данные
			ASN1.ISO.PKCS.PKCS7.EncapsulatedContentInfo encapContentInfo = 
				new ASN1.ISO.PKCS.PKCS7.EncapsulatedContentInfo(
					new ASN1.ObjectIdentifier(dataType), 
                    new ASN1.OctetString(authenticatedSafe.Encoded)
			); 
			// подписать данные
			ASN1.ISO.PKCS.PKCS7.SignerInfo signerInfo = CMS.SignData(
				rand, privateKey, certificate, hashParameters, signParameters, 
                encapContentInfo, authAttributes, unauthAttributes
			); 
			// закодировать зашифрованные данные
			ASN1.ISO.PKCS.PKCS7.SignerInfos signerInfos = 
				new ASN1.ISO.PKCS.PKCS7.SignerInfos(
					new ASN1.ISO.PKCS.PKCS7.SignerInfo[] {signerInfo}
			);
			// закодировать структуру CMS
			ASN1.ISO.PKCS.PKCS7.SignedData signedData = 
				new ASN1.ISO.PKCS.PKCS7.SignedData(version, 
					digestAlgorithms, encapContentInfo, certificates, crls, signerInfos
			); 
			// закодировать данные
			ASN1.ISO.PKCS.ContentInfo contentInfo = new ASN1.ISO.PKCS.ContentInfo(
				new ASN1.ObjectIdentifier(ASN1.ISO.PKCS.PKCS7.OID.signedData), signedData
			); 
			// вернуть созданный контейнер
			return new ASN1.ISO.PKCS.PKCS12.PFX(new ASN1.Integer(3), contentInfo, null); 
		}
		///////////////////////////////////////////////////////////////////////
		// Создать контейнер PKCS12 с вычисленной имитовставкой
		///////////////////////////////////////////////////////////////////////
		public static ASN1.ISO.PKCS.PKCS12.PFX CreateAuthenticatedContainer(
			Factory factory, ASN1.ISO.PKCS.PKCS12.AuthenticatedSafe authenticatedSafe, 
			ASN1.ISO.AlgorithmIdentifier hashParameters, byte[] salt, int iterations, string password)
		{
			// закодировать данные
			byte[] encoded = authenticatedSafe.Encoded;  

			// создать алгоритм хэширования 
			using (Hash hashAlgorithm = factory.CreateAlgorithm<Hash>(null, hashParameters))
            { 
			    // при ошибке выбросить исключение
			    if (hashAlgorithm == null) throw new NotSupportedException(); 

			    // создать алгоритм вычисления имитовставки
			    using (Mac macAlgorithm = new PBE.PBMACP12(hashAlgorithm, salt, iterations)) 
                {
			        // закодировать пароль
			        using (ISecretKey encodedPassword = SecretKey.FromPassword(password, Encoding.UTF8))
                    { 
			            // вычислить имитовставку от данных
			            ASN1.OctetString mac = new ASN1.OctetString(macAlgorithm.MacData(
				            encodedPassword, encoded, 0, encoded.Length
			            )); 
			            // закодировать данные
			            ASN1.ISO.PKCS.ContentInfo contentInfo = new ASN1.ISO.PKCS.ContentInfo(
				            new ASN1.ObjectIdentifier(ASN1.ISO.PKCS.PKCS7.OID.data), 
                            new ASN1.OctetString(encoded)
			            ); 
			            // закодировать имитовставку от данных
			            ASN1.ISO.PKCS.DigestInfo digestInfo = 
				            new ASN1.ISO.PKCS.DigestInfo(hashParameters, mac); 

			            // закодировать информацию об имитовставке
			            ASN1.ISO.PKCS.PKCS12.MacData macData = new ASN1.ISO.PKCS.PKCS12.MacData(
				            digestInfo, new ASN1.OctetString(salt), new ASN1.Integer(iterations)
			            ); 
			            // вернуть созданную структуру
			            return new ASN1.ISO.PKCS.PKCS12.PFX(new ASN1.Integer(3), contentInfo, macData); 
                    }
                }
            }
		}
		///////////////////////////////////////////////////////////////////////
		// Проверить имитовставку контейнера
		///////////////////////////////////////////////////////////////////////
		public static void CheckAuthenticatedContainer(Factory factory, 
			ASN1.ISO.PKCS.PKCS12.PFX content, string password)
		{
			// получить закодированные данные
			byte[] encoded = content.GetAuthSafeContent().Encoded;

			// получить закодированную имитовставку
			ASN1.ISO.PKCS.PKCS12.MacData macData = content.MacData; 

            // получить salt-значение
            byte[] salt = macData.MacSalt.Value; 

		    // получить число итераций
		    int iterations = macData.Iterations.Value.IntValue; 

            // получить имитовставку
            byte[] mac = macData.Mac.Digest.Value; 

			// создать алгоритм хэширования 
			using (Hash hashAlgorithm = factory.CreateAlgorithm<Hash>(null, macData.Mac.DigestAlgorithm))
            { 
			    // при ошибке выбросить исключение
			    if (hashAlgorithm == null) throw new NotSupportedException(); 

			    // создать алгоритм вычисления имитовставки
			    using (Mac macAlgorithm = new PBE.PBMACP12(hashAlgorithm, salt, iterations))
                {
			        // закодировать пароль
			        using (ISecretKey encodedPassword = SecretKey.FromPassword(password, Encoding.UTF8))
                    { 
			            // вычислить имитовставку от данных
			            byte[] check = macAlgorithm.MacData(encodedPassword, encoded, 0, encoded.Length); 

			            // извлечь имитовставку для сравнения
			            if (!Arrays.Equals(check, mac)) 
                        {
                            // создать алгоритм вычисления имитовставки
                            using (Mac macAlgorithm2 = new PBE.PBMACTC26(hashAlgorithm, salt, iterations))
                            {
                                // вычислить имитовставку от данных
                                check = macAlgorithm2.MacData(encodedPassword, encoded, 0, encoded.Length);
                            }
                        }
                        // проверить совпадение имитовставок
                        if (!Arrays.Equals(check, mac)) throw new InvalidCredentialException();
                    }
                }
            }
		}
		///////////////////////////////////////////////////////////////////////
		// Зашифровать элемент 
		///////////////////////////////////////////////////////////////////////
		internal static ASN1.ISO.PKCS.PKCS12.SafeBag Encrypt(PfxData<ASN1.ISO.PKCS.PKCS12.SafeBag> safeBag)
		{
			// проверить наличие шифрования
			if (safeBag.Encryptor == null) return safeBag.Content; 

			// в зависимости от типа
			if (safeBag.Content.BagId.Value == ASN1.ISO.PKCS.PKCS12.OID.bt_key)
			{
				// раскодировать тип данных
				ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo keyInfo = 
					new ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo(safeBag.Content.BagValue);

				// зашифровать данные
				ASN1.IEncodable encodable = ASN1.Encodable.Decode(
                    safeBag.Encryptor.Encrypt(keyInfo.Encoded)
                ); 
				// преобразовать тип данных
				ASN1.ISO.PKCS.PKCS8.EncryptedPrivateKeyInfo encryptedKey = 
					new ASN1.ISO.PKCS.PKCS8.EncryptedPrivateKeyInfo(encodable);

				// вернуть зашифрованное представление
				return new ASN1.ISO.PKCS.PKCS12.SafeBag(
					new ASN1.ObjectIdentifier(ASN1.ISO.PKCS.PKCS12.OID.bt_shroudedKey), 
                    encryptedKey, safeBag.Content.BagAttributes);
			}
			else return safeBag.Content; 
		}
		///////////////////////////////////////////////////////////////////////
		// Расшифровать элемент
		///////////////////////////////////////////////////////////////////////
		internal static PfxData<ASN1.ISO.PKCS.PKCS12.SafeBag> Decrypt(
            ASN1.ISO.PKCS.PKCS12.SafeBag safeBag, PfxDecryptor decryptor)
		{
			// для незашифрованного элемента
			if (safeBag.BagId.Value != ASN1.ISO.PKCS.PKCS12.OID.bt_shroudedKey)
			{
			    // вернуть исходное содержимое
			    return new PfxData<ASN1.ISO.PKCS.PKCS12.SafeBag>(safeBag, null);
            }
			// расшифровать данные
			PfxData<byte[]> decrypted = decryptor.Decrypt(safeBag.BagValue.Encoded,
                typeof(ASN1.ISO.PKCS.PKCS8.EncryptedPrivateKeyInfo)
            );
			// преобразовать тип данных 
			ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo keyInfo = 
				new ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo(
                    ASN1.Encodable.Decode(decrypted.Content)
            );
			// сформировать элемент контейнера
			safeBag = new ASN1.ISO.PKCS.PKCS12.SafeBag(
    			new ASN1.ObjectIdentifier(ASN1.ISO.PKCS.PKCS12.OID.bt_key), 
                keyInfo, safeBag.BagAttributes
            );
            // вернуть элемент контейнера
            return new PfxData<ASN1.ISO.PKCS.PKCS12.SafeBag>(
                safeBag, decrypted.Encryptor
            ); 
		}
	}
}
